src-ghc/Pact/GasModel/GasModel.hs

```diff
@@ -0,0 +1,331 @@
+newtype Average = Average Double
+instance Csv.ToField Average where
+  toField (Average avg) = Csv.toField avg
+
+average :: [Double] -> Average
+average nums = Average avg
+  where meanSum = foldl' (+) 0.0 nums
+        total = fromIntegral (length nums)
+        avg = meanSum / total
+
+
+newtype GasPrice = GasPrice Integer
+instance Csv.ToField GasPrice where
+  toField (GasPrice p) = Csv.toField p
+
+gasPriceDivisor :: Double
+gasPriceDivisor = 2500.0
+
+gasPrice :: [NanoSeconds] -> GasPrice
+gasPrice measurements = GasPrice price
+  where
+    (Average avg) = average measurements
+    price = ceiling (avg / gasPriceDivisor)
+
+
+exec
+  :: EvalState
+  -> EvalEnv e
+  -> [Term Name]
+  -> IO (Either PactError [Term Name], EvalState)
+exec s e terms = runEval' s e $ mapM eval terms
+
+
+mockRun
+  :: PactExpression
+  -> (EvalEnv e, EvalState)
+  -> IO (Either PactError [Term Name], EvalState)
+mockRun expr (env, state) = do
+  terms <- compileCode (_pactExpressionFull expr)
+  exec state env terms
+
+
+setupEnv :: GasSetup e -> IO (EvalEnv e, EvalState)
+setupEnv (GasSetup e s _ _) = do
+  env <- e
+  state <- s
+  return (env, state)
+
+
+mockRuns
+  :: GasUnitTests
+  -> IO ()
+mockRuns tests = do
+  void $ mapOverGasUnitTests tests run run
+  where
+    run expr dbSetup = do
+      putStrLn $ "Dry run for " ++ T.unpack (getDescription expr dbSetup)
+
+      (res,_) <- bracket (setupEnv dbSetup)
+                         (gasSetupCleanup dbSetup)
+                         (mockRun expr)
+      eitherDie (getDescription expr dbSetup) res
+
+
+-- | For debugging purposes
+mockRuns'
+  :: GasUnitTests
+  -> IO [GasTestResult (Either PactError [Term Name], EvalState)]
+mockRuns' tests = do
+  mapOverGasUnitTests tests run run
+  where
+    run expr dbSetup = do
+      bracket (setupEnv dbSetup) (gasSetupCleanup dbSetup) (mockRun expr)
+
+
+bench
+  :: PactExpression
+  -> GasSetup e
+  -> IO NanoSeconds
+bench expr dbSetup = do
+  terms <- compileCode (_pactExpressionFull expr)
+  putStrLn $ T.unpack (getDescription expr dbSetup)
+  report <- bracket setup teardown $ C.benchmark' . (run terms)
+  return $ secToNs $
+           estPoint $
+           C.anMean $
+           C.reportAnalysis report
+  where
+    setup = do
+      s <- setupEnv dbSetup
+      return $ NoopNFData s
+    teardown (NoopNFData env) = do
+      (gasSetupCleanup dbSetup) env
+    run terms ~(NoopNFData (env, state)) =
+          C.nfIO (exec state env terms)
+
+
+benchesOnce
+  :: GasUnitTests
+  -> IO ()
+benchesOnce tests =
+  void $ mapOverGasUnitTests tests bench bench
+
+
+type Means = (NanoSeconds, NanoSeconds, NanoSeconds)
+
+-- | Benchmarks each native function's tests three times
+benchesMultiple
+  :: (NativeDefName, GasUnitTests)
+  -> IO (NativeDefName, [GasTestResult Means])
+benchesMultiple (funName, tests) = do
+  results <- mapOverGasUnitTests tests run run
+  return (funName, results)
+  where
+    run expr setup = do
+      mean1 <- bench expr setup
```

Comment:
Instead of fixing the number of trials at three, it might be better to generalize to `n` so you can vary the batch size --

```
    means <- replicateM n $ bench expr setup
    return $! average means    -- or return the samples themselves
```